README.md
# [Motio](http://darsa.in/motio)

Small JavaScript library for simple but powerful sprite based animations and panning.

Motio takes an element and animates its background position to create an animation effect. All is super optimized for
speed and chained to the requestAnimationFrame, with a polyfill for older browsers without it.

That being said, animating the element background is not the fastest possible way how to do this (canvas solutions are
a lot faster), but it is sure as hell the most simple one, and compatible with everything from IE6 and up.

#### Dependencies

Motio has no dependencies, but there is an optional
[Motio jQuery plugin extension](https://raw.github.com/darsain/motio/master/dist/jquery.motio.min.js) available.

#### Compatibility

Works everywhere.

## Usage

Sprite animation mode:

```js
var element = document.querySelector('#sprite');
var sprite = new Motio(element, {
	fps: 10,
	frames: 14
});
sprite.play();   // start animation
sprite.pause();  // pause animation
sprite.toggle(); // toggle play/pause
```

Seamless background panning mode:

```js
var element = document.querySelector('#panning');
var panning = new Motio(element, {
	fps: 30, // Frames per second. More fps = higher CPU load.
	speedX: -30 // Negative horizontal speed = panning to left.
});
sprite.play();    // start animation
sprite.pause();   // pause animation
sprite.toggle();  // toggle play/pause
sprite.toStart(); // animate to 1st frame and stop
sprite.toEnd();   // animate to last frame and stop
sprite.to(10);    // animate to 11th frame and stop
```

## Download

Latest stable release:

- [Production `motio.min.js`](https://raw.github.com/darsain/motio/master/dist/motio.min.js) - 3KB, 1.6KB gzipped
- [Development `motio.js`](https://raw.github.com/darsain/motio/master/dist/motio.js) - 12KB

jQuery plugin version:

- [Production `jquery.motio.min.js`](https://raw.github.com/darsain/motio/master/dist/jquery.motio.min.js) - 3.7KB, 1.8KB gzipped
- [Development `jquery.motio.js`](https://raw.github.com/darsain/motio/master/dist/jquery.motio.js) - 13.5KB

When isolating issues on jsfiddle, you can use this URL:

- [http://darsain.github.io/motio/js/motio.min.js](http://darsain.github.io/motio/js/motio.min.js)

## Documentation

Can be found in the [docs](https://github.com/darsain/motio/tree/master/docs) directory.

## Contributing

Please, read the [Contributing Guidelines](CONTRIBUTING.md) for this project.

CHANGELOG.md
**2.2.2** :: *7th Aug 2014*

- Fixing another extremely rare jamming of animations. Rendering pipeline has been more streamlined so there shouldn't be more of this.

**2.2.1** :: *18th Apr 2013*

- Fixing extremely rare jamming of sprite based animations.

**2.2.0** :: *11th Apr 2013*

- Exposing `element` property.
- Fixing immediate animation breaking other animations.

**2.1.0** :: *6th Apr 2013*

- Adding `width` & `height` options.

**2.0.2** :: *5th Apr 2013*

- Removing forgotten toggle event.

**2.0.1** :: *5th Apr 2013*

- Fixing `.to()` method not returning current instance.

**2.0.0** :: *5th Apr 2013*

- Dropping jQuery dependency.
- Adding optional jQuery plugin version of Motio.
- Improved events API.
- Method `.toStart()` - when called on an animation that is already at the start - will repeat the animation from the end. Ditto for the `.toEnd()` method.
- Methods `.toStart()`, `.toEnd()`, and `.to()` now accept callbacks.
- Removed `toStart`, `toEnd`, and `to` events.
- Added reversed sprite based animation playback by passing `true` into the `.play()` method's first argument.

**1.0.0** :: *25th May 2012*

The initial Motio version, dependent on jQuery.

js/main.js
/*global Motio */
jQuery(function ($) {
	'use strict';

	var windowSpy = new $.Espy(window);

	// ==========================================================================
	//   Header clouds
	// ==========================================================================
	(function () {
		var header = $('header')[0];
		var headerClouds = new Motio(header, {
			fps: 30,
			speedX: 60,
			bgWidth: 1024,
			bgHeight: 1024
		});

		// Play only when in the viewport
		windowSpy.add(header, function (entered) {
			headerClouds[entered ? 'play' : 'pause']();
		});
	}());

	// ==========================================================================
	//   Examples
	// ==========================================================================

	// Panning
	(function () {
		var $example = $('#panning .example');
		var frame = $example.find('.frame')[0];
		var offset = $example.offset();
		var motio = new Motio(frame, {
			fps: 30,
			bgWidth: 1024,
			bgHeight: 1024
		});

		// Play/Pause when mouse enters/leaves the frame
		$example.on('mouseenter mouseleave', function (event) {
			if (event.type === 'mouseenter') {
				motio.play();
			} else {
				motio.pause();
			}
		});

		// Update example offset offset on window resize
		$(window).on('resize', function () {
			offset = $example.offset();
		});

		// Update the animation speed & direction based on a cursor position
		$example.on('mousemove', function (event) {
			motio.set('speedX', event.pageX - offset.left - motio.width / 2);
			motio.set('speedY', event.pageY - offset.top - motio.height / 2);
		});
	}());

	// Sprite
	$('a[data-toggle="tab"][href=#sprite]').one('shown', function () {
		var $example = $('#sprite .example');
		var frame = $example.find('.frame')[0];
		var motio = new Motio(frame, {
			fps: 10,
			frames: 14
		});

		// Play when mouse enters the frame, and pause when it leaves
		$example.on('mouseenter mouseleave', function (event) {
			motio[event.type === 'mouseenter' ? 'play' : 'pause']();
		});
	});

	// 360 view
	$('a[data-toggle="tab"][href=#circview]').one('shown', function () {
		var $example = $('#circview .example');
		var exampleLeft = $example.offset().left;
		var exampleWidth = $example.width();
		var frame = $example.find('.frame')[0];
		var motio = new Motio(frame, {
			frames: 18
		});

		// Update example left offset on window resize
		$(window).on('resize', function () {
			exampleLeft = $example.offset().left;
		});

		// Activate frame based on the cursor position
		$example.on('mousemove', function (event) {
			motio.to(Math.floor(motio.frames / exampleWidth * (event.pageX - exampleLeft)), true);
		});
	});

	// Extreme spriting - minigame
	// This is really dumb, and on top of that adapted from older version of Motio, so please ignore.
	$('a[data-toggle="tab"][href=#game]').one('shown', function () {
		var $game = $('#game .frame');
		var pos = 350;
		var $char = $game.find('.char').css({ left: pos + 'px' });
		var posMax = $game.innerWidth() - $char.innerWidth();
		var facing = 'right';
		var moveSpeed = 300;
		var moveFps = 30;
		var pressed = [];
		var inAction = 0;
		var isRunning = 0;
		var mIndex;
		var listenOn = [37,39,32,66];
		var $mations = $char.children();
		var mations = {
			right: {
				stand: $mations.filter('.stand').motio({ frames: 8, startPaused: 1, fps: 10 }),
				run:   $mations.filter('.run'  ).motio({ frames: 6, startPaused: 1, fps: 10 }),
				jump:  $mations.filter('.jump' ).motio({ frames: 10, startPaused: 1, fps: 15 }),
				kick:  $mations.filter('.kick' ).motio({ frames: 9, startPaused: 1, fps: 15 })
			},
			left: {
				stand: $mations.filter('.stand_left').motio({ frames: 8, startPaused: 1, fps: 10 }),
				run:   $mations.filter('.run_left'  ).motio({ frames: 6, startPaused: 1, fps: 10 }),
				jump:  $mations.filter('.jump_left' ).motio({ frames: 10, startPaused: 1, fps: 15 }),
				kick:  $mations.filter('.kick_left' ).motio({ frames: 9, startPaused: 1, fps: 15 })
			}
		};

		// Hide everything on start
		$mations.hide();

		// Start with standing animation
		mations[facing].stand.show().motio('play');

		// Resets the stance back to running or standing after actions like kick
		function resetStance() {
			/*jshint validthis:true */
			inAction = 0;
			$(this.element).hide();
			mations[facing][isRunning ? 'run' : 'stand'].show().motio('play');
		}

		// Keydown handlers
		$(document).on('keydown', function (event) {
			if ($.inArray(event.which, listenOn) === -1 || pressed[event.which]) {
				return;
			}
			pressed[event.which] = true;

			var request;
			switch (event.which) {
				// Left arrow
				case 37:
					request = 'run';
					facing = 'left';
					break;

				// Right arrow
				case 39:
					request = 'run';
					facing = 'right';
					break;

				// Spacebar
				case 32:
					request = 'jump';
					break;

				// B
				case 66:
					request = 'kick';
					break;
			}

			// Show concerned animation
			$mations.hide().motio('toStart', true);
			mations[facing][request].show();

			if (request === 'run') {
				inAction = 0;
				mIndex = clearTimeout(mIndex);
				isRunning = 1;
				move();
				mations[facing][request].motio('play');
			} else {
				inAction = 1;
				mations[facing][request].motio('toEnd', resetStance);
			}

			return false;
		});

		// Keyup handlers
		$(document).on('keyup', function (event) {
			if ($.inArray(event.which, listenOn) === -1) {
				return;
			}
			pressed[event.which] = false;

			var released;
			switch (event.which) {
				// Left & arrow
				case 37:
					released = 'left';
					break;

				// Right arrow
				case 39:
					released = 'right';
					break;
			}

			if (isRunning && facing === released) {
				mations[released].run.hide().motio('toStart', true);
				isRunning = 0;
				mIndex = clearTimeout(mIndex);

				if (!inAction) {
					mations[facing].stand.show().motio('play');
				}
			}
			return false;
		});

		// Move function
		function move() {
			if (pos === 0 && facing === 'left' || pos === posMax && facing === 'right') {
				return;
			}

			pos += (facing === 'right' ? moveSpeed : -moveSpeed) / moveFps;

			if (pos < 0) {
				pos = 0;
			} else if (pos > posMax) {
				pos = posMax;
			}

			$char[0].style.left = pos + 'px';
			mIndex = setTimeout(move, 1000 / moveFps);
		}
	});
});

compress.sh
#/bin/sh
#
# @description : BASH script for minifying javascript files
# @requirements : curl

# Format for minifying multiple files:
# IN=( file1.js file2.js file3.js )
IN=jquery.motio.js
OUT=jquery.motio.min.js

# Compiler settings
API_URL=http://closure-compiler.appspot.com/compile
COMPILATION_LEVEL=SIMPLE_OPTIMIZATIONS

# Check if curl is installed
if [ -z "$(which curl)" ]
then
	echo 'Please install curl to proceed.'
	exit
fi

# Itearate through all files
for f in ${IN[@]}
do
	if [ -r ${f} ]
	then
		code="${code} --data-urlencode js_code@${f}"
	else
		echo "File ${f} does not exist or is not readable. Skipped."
	fi
done

# If there is no code, terminate
if [ -z "${code}" ]
then
	echo 'Nothing to compile.'
	exit
fi

# Compile & save new file
`curl \
	--url ${API_URL} \
	--header 'Content-type: application/x-www-form-urlencoded' \
	${code} \
	--data output_format=text \
	--data output_info=compiled_code \
	--data compilation_level=${COMPILATION_LEVEL} \
	--output ${OUT}`


Gruntfile.js
/*jshint node:true */
module.exports = function (grunt) {
	'use strict';

	// Override environment based line endings enforced by Grunt
	grunt.util.linefeed = '\n';

	// Grunt configuration
	grunt.initConfig({
		pkg: grunt.file.readJSON('meta.json'),
		meta: {
			banner: '/*!\n' +
				' * <%= pkg.name %> <%= pkg.version %> - <%= grunt.template.today("dS mmm yyyy") %>\n' +
				' * <%= pkg.repository.homepage %>\n' +
				' *\n' +
				' * Licensed under the <%= pkg.licenses[0].type %> license.\n' +
				' * <%= pkg.licenses[0].url %>\n' +
				' */\n',
			bannerLight: '/*! <%= pkg.name %> <%= pkg.version %>' +
				' - <%= grunt.template.today("dS mmm yyyy") %> | <%= pkg.repository.homepage %> */'
		},

		// JSHint the code.
		jshint: {
			options: {
				jshintrc: true
			},
			all: ['src/*.js']
		},

		// Clean folders.
		clean: {
			dist: ['dist/**', '!dist']
		},

		// Concatenate files.
		concat: {
			options: {
				banner: '<%= meta.banner %>'
			},
			vanilla: {
				src: 'src/<%= pkg.name %>.js',
				dest: 'dist/<%= pkg.name %>.js'
			},
			jquery: {
				src: ['src/<%= pkg.name %>.js', 'src/jquery.js'],
				dest: 'dist/jquery.<%= pkg.name %>.js'
			}
		},

		// Minify with Google Closure Compiler.
		gcc: {
			options: {
				banner: '<%= meta.bannerLight %>'
			},
			vanilla: {
				src: 'src/<%= pkg.name %>.js',
				dest: 'dist/<%= pkg.name %>.min.js'
			},
			jquery: {
				src: ['src/<%= pkg.name %>.js', 'src/jquery.js'],
				dest: 'dist/jquery.<%= pkg.name %>.min.js'
			}
		},

		// Bump up fields in JSON files.
		bumpup: {
			options: {
				updateProps: {
					pkg: 'meta.json',
				},
			},
			files: ['meta.json', '<%= pkg.name %>.jquery.json'],
		},

		// Commit changes and tag the latest commit with a version from JSON file.
		tagrelease: '<%= pkg.version %>'
	});

	// These plugins provide necessary tasks.
	grunt.loadNpmTasks('grunt-contrib-jshint');
	grunt.loadNpmTasks('grunt-contrib-concat');
	grunt.loadNpmTasks('grunt-contrib-clean');
	grunt.loadNpmTasks('grunt-tagrelease');
	grunt.loadNpmTasks('grunt-bumpup');
	grunt.loadNpmTasks('grunt-gcc');

	// Build task.
	grunt.registerTask('build', function () {
		grunt.task.run('clean');
		grunt.task.run('concat');
		grunt.task.run('gcc');
	});

	// Release task.
	grunt.registerTask('release', function (type) {
		type = type ? type : 'patch';
		grunt.task.run('jshint');
		grunt.task.run('bumpup:' + type);
		grunt.task.run('build');
		grunt.task.run('tagrelease');
	});

	// Default task.
	grunt.registerTask('default', ['jshint']);
};

src/jquery.js
/*global Motio */
(function ($) {
	'use strict';

	// Names
	var pluginName = 'motio';
	var namespace  = pluginName;

	// jQuery plugin
	$.fn[pluginName] = function (options, callbackMap) {
		var method, methodArgs;

		// Attributes logic
		if (!$.isPlainObject(options)) {
			if (typeof options === 'string' || options === false) {
				method = options === false ? 'destroy' : options;
				methodArgs = Array.prototype.slice.call(arguments, 1);
			}
			options = {};
		}

		// Apply plugin to all elements
		return this.each(function (i, element) {
			// Plugin call with prevention against multiple instantiations
			var plugin = $.data(element, namespace);

			if (!plugin && !method) {
				// Create a new plugin object if it doesn't exist yet
				plugin = $.data(element, namespace, new Motio(element, options));
				// Bind callbacks
				plugin.on(callbackMap);
				// Start playing when requested
				if (!options.startPaused) {
					plugin.play();
				}
			} else if (plugin && method) {
				// Call plugin method
				if (plugin[method]) {
					plugin[method].apply(plugin, methodArgs);
				}
				// Remove plugin from element data on destroy
				if (method === 'destroy') {
					$.removeData(element, namespace);
				}
			}
		});
	};
}(jQuery));

jquery.motio.js
/*!
 * jQuery Motio v1.0.1
 * https://github.com/Darsain/motio
 *
 * Licensed under the MIT license.
 * http://www.opensource.org/licenses/MIT
 */

/*jshint eqeqeq: true, noempty: true, strict: true, undef: true, expr: true, smarttabs: true, browser: true */
/*global jQuery:false */

;(function($, undefined){
'use strict';

	var pluginName = 'motio',
		namespace = 'plugin_' + pluginName,
		cAF = window.cancelAnimationFrame || window.cancelRequestAnimationFrame,
		rAF = window.requestAnimationFrame;

	// Plugin "class"
	function Plugin( frame, options ){

		// Private global variables
		var $frame, o, isPan, tIndex, lastTime, frameSize, frames, active, originalBgPos, bgPos, affected, type, to,
			self = this,
			callbacks = !$.Callbacks ? false : {};

		/**
		 * Pause animation
		 *
		 * @public
		 */
		this.pause = function(){

			if( tIndex ){

				tIndex = clearTimeout( tIndex );

				trigger( 'pause' );

			}

		};

		/**
		 * Play animation
		 *
		 * @public
		 */
		this.play = function(){

			if( !tIndex ){

				render();
				trigger( 'play' );

			}

		};

		/**
		 * Pause or Play animation, depending on its current state
		 *
		 * @public
		 */
		this.toggle = function(){

			trigger( 'toggle' );
			self[ tIndex ? 'pause' : 'play' ]();

		};

		/**
		 * Update one of the dynamic option properties
		 *
		 * Only some options can be updated:
		 *  speed, fps
		 *
		 * @public
		 *
		 * @param {String} option Option name
		 * @param {Mixed} value   New option value
		 */
		this.set = function( option, value ){

			var updateable = [ 'speed', 'fps' ];

			if( $.inArray( option, updateable ) !== -1 ){

				o[option] = value;

			}

		};

		/**
		 * Animate or directly set sprite to its first frame and than pause
		 *
		 * @public
		 *
		 * @param {Bool} immediate Move to start immediately without animation
		 */
		this.toStart = function( immediate ){

			if( isPan ){

				return;

			}

			if( immediate || active === 0 ){

				type = false;

				trigger( 'start' );
				self.pause();
				setPos( 0 );

			} else {

				type = 'toStart';

				self.play();

			}

		};


		/**
		 * Animate or directly set sprite to its last frame and than pause
		 *
		 * @public
		 *
		 * @param {Bool} immediate Move to end immediately without animation
		 */
		this.toEnd = function( immediate ){

			if( isPan ){

				return;

			}

			if( immediate || active === frames.length - 1 ){

				type = 0;

				trigger( 'end' );
				self.pause();
				setPos( frames.length -1 );

			} else {

				type = 'toEnd';

				self.play();

			}

		};


		/**
		 * Animate or directly set sprite to passed frame index and than pause
		 *
		 * @public
		 *
		 * @param {Int}  frame     Frame index starting at 0
		 * @param {Bool} immediate Move to end immediately without animation
		 */
		this.to = function( frame, immediate ){

			if( isPan || !isNumber( frame ) || frame < 0 && frame >= frames.length ){

				return;

			}

			if( immediate || frame === active ){

				type = 0;

				trigger( 'to', [ frame ] );
				self.pause();
				setPos( frame );

			} else {

				type = 'to';
				to = frame;

				self.play();

			}

		};

		/**
		 * Binds callbacks to custom event lists using jQuery.Callbacks
		 *
		 * @public
		 *
		 * @param  {String}   eName Event name
		 * @param  {Function} fn    Callback function, or array with functions
		 */
		this.on = function( eName, fn ){

			if( callbacks && fn ){

				if( !callbacks[eName] ){

					callbacks[eName] = $.Callbacks('unique');

				}

				callbacks[eName].add( fn );

			}

		};

		/**
		 * Remove on or all callbacks from custom event list
		 *
		 * @public
		 *
		 * @param  {String}   eName Event name
		 * @param  {Function} fn    Callback function to be removed. If nothing is passed, all callbacks for 'eName' will be removed
		 */
		this.off = function( eName, fn ){

			if( callbacks && callbacks[eName] ){

				if( fn ){

					callbacks[eName].remove( fn );

				} else {

					callbacks[eName].empty();

				}

			}

		};

		/**
		 * Trigger callbacks from custom event list
		 *
		 * @public
		 *
		 * @param  {String} eName Event name
		 * @param  {Array}  args  Array with arguments for callback functions
		 */
		function trigger( eName, args ){

			if( callbacks && callbacks[eName] ){

				callbacks[eName].fireWith( frame, args );

			}

			$frame.trigger( pluginName + ':' + eName, args );

		}

		/**
		 * Destroy plugin instance and reset backgroundPosition to its original state
		 *
		 * @public
		 */
		this.destroy = function(){

			self.pause();
			$frame.css('backgroundPosition', originalBgPos);
			$.removeData( frame, namespace );

		};

		/**
		 * Render animation frame
		 *
		 * @public
		 */
		function render(){

			var newpos;

			// Call next frame
			tIndex = setTimeout( function(){

				rAF( render );

			}, 1000 / o.fps );


			if( isPan ){

				bgPos[affected] = bgPos[affected] + ( o.speed / o.fps );

				if( o.bgSize > 0 && Math.abs( bgPos[affected] ) > o.bgSize ){

					bgPos[affected] = bgPos[affected] % o.bgSize;

				}

				newpos = bgPos;

			} else {

				switch( type ){

					case 'toStart':

						if( --active <= 0 ){

							type = active = 0;

							self.pause();

							trigger( 'start' );

						}

					break;

					case 'toEnd':

						if( ++active >= frames.length -1 ){

							type = 0;
							active = frames.length - 1;

							self.pause();

							trigger( 'end' );

						}

					break;

					case 'to':

						if( active < to ){

							active++;

						} else if( active > to ){

							active--;

						}

						if( active === to ){

							trigger( 'to', [ frame ] );
							self.pause();

						}

					break;

					default:

						if( ++active >= frames.length ){

							active = 0;

						}

				}

				newpos = active;

			}

			// Update background position
			setPos( newpos );

		}


		/**
		 * Set frame position
		 *
		 * @param {Mixed} arg Frame index in 'frames' array, or position object
		 */
		function setPos( arg ){

			var pos = !isPan && isNumber( arg ) ? frames[arg] : typeof arg === 'object' ? arg.x + 'px ' + arg.y + 'px' : false;

			if( pos ){

				frame.style.backgroundPosition = pos;

				if( !isPan ){

					active = arg;

				}

				trigger( 'frame', isPan ? [ arg[affected], o.bgSize ] : [ arg, frames.length ] );

			}

		}


		/**
		 * Check whether the value is a number
		 *
		 * @private
		 *
		 * @param  {Mixed}  value Value to be checked
		 *
		 * @return {Boolean} True if number, false if not
		 */
		function isNumber( value ){

			return !isNaN( parseFloat( value ) ) && isFinite( value );

		}


		/** Construct */
		(function(){

			// Set variables
			$frame    = $(frame);
			o         = $.extend( {}, $.fn[pluginName].defaults, options );
			isPan     = !o.frames;
			tIndex    = 0;
			frameSize = $frame[ o.vertical ? 'innerHeight' : 'innerWidth' ]();
			affected  = o.vertical ? 'y' : 'x';
			originalBgPos = $frame.css('backgroundPosition') || $frame.css('backgroundPositionX') + ' ' + $frame.css('backgroundPositionY');

			// Background position
			var posString = originalBgPos.replace(/left|top/g, 0).split(" ");
			bgPos = {
				x: parseInt( posString[0], 10 ),
				y: parseInt( posString[1], 10 )
			};

			// Build frames array
			if( !isPan ){

				var tmpPos = bgPos;
				frames = [];

				for( var i = 0; i < o.frames; i++ ){

					tmpPos[affected] = i * -frameSize;

					frames.push( tmpPos.x + 'px ' + tmpPos.y + 'px' );

				}

				active = -1;

			}

			// Start animation
			if( !o.paused ){

				lastTime = 0;
				render();

			}

		}());

	}


	// jQuery plugin extension
	$.fn[pluginName] = function( options, returnInstance ){

		var method = false,
			methodArgs,
			instances = [];

		// Basic attributes logic
		if( typeof options !== 'undefined' && !$.isPlainObject( options ) ){
			method = options === false ? 'destroy' : options;
			methodArgs = arguments;
			Array.prototype.shift.call( methodArgs );
		}

		// Apply requested actions on all elements
		this.each(function( i, e ){

			// Plugin call with prevention against multiple instantiations
			var plugin = $.data( e, namespace );

			if( plugin && method ){

				// Call plugin method
				if( plugin[method] ){

					plugin[method].apply( plugin, methodArgs );

				}

			} else if( !plugin && !method ){

				// Create a new plugin object if it doesn't exist yet
				plugin =  $.data( e, namespace, new Plugin( e, options ) );

			}

			// Push plugin to instances
			instances.push( plugin );

		});

		// Return chainable jQuery object, or plugin instance(s)
		return returnInstance && !method ? instances.length > 1 ? instances : instances[0] : this;

	};


	// Default options
	$.fn[pluginName].defaults = {
		// global options
		fps:      15,    // frames per second - bigger number means smoother animations but higher CPU load
		vertical: 0,     // true for vertical sprites
		paused:   0,     // whether to start motio paused

		// sprite based animation specific options ("frames === 0" means that you are requesting panning animation)
		frames:   0,     // number of frames in sprite

		// pan specific options
		speed:    50,    // number of pixels to move per second (use negative number to go in an opposite direction)
		bgSize:   0      // size of the background image in animated direction (width for horizontal, height for vertical)
						 // it is needed so the script will know when to reset the background position to 0, and thus not overflow the JavaScript 2^53 integer limit
						 // when 0 (=unknown), the position will iterate into ridiculous numbers, which might eventually result into a buggy animation later on...
	};


	// local requestAnimationFrame polyfill
	(function (w) {
		var vendors = ['ms', 'moz', 'webkit', 'o'],
			lastTime = 0;

		// For a more accurate WindowAnimationTiming interface implementation, ditch the native
		// requestAnimationFrame when cancelAnimationFrame is not present (older versions of Firefox)
		for(var x = 0; x < vendors.length && !cAF; ++x) {
			cAF = w[vendors[x]+'CancelAnimationFrame'] || w[vendors[x]+'CancelRequestAnimationFrame'];
			rAF = cAF && w[vendors[x]+'RequestAnimationFrame'];
		}

		if (!cAF) {
			rAF = function (callback) {
				var currTime = +new Date(),
					timeToCall = Math.max(0, 16 - (currTime - lastTime));
				lastTime = currTime + timeToCall;
				return w.setTimeout(function () { callback(currTime + timeToCall); }, timeToCall);
			};

			cAF = function (id) {
				clearTimeout(id);
			};
		}
	}(window));

})(jQuery);